### ISSUE TITLE
VAE performance is bad (with standard parameters)

### ISSUE THREAD

**Issue by Faur** [AppliedDataSciencePartners/WorldModels#12]
Running the code as is I get the following results after training the VAE (`python 02_train_vae.py --start_batch 0 --max_batch 9 --new_model`):

![image](https://user-images.githubusercontent.com/8115763/42577811-7f2c1398-8525-11e8-94cb-9c6ec76ceb03.png)

> `loss: 0.1416 - vae_r_loss: 0.1416 - vae_kl_loss: 3.8347e-07 - val_loss: 0.1378 - val_vae_r_loss: 0.1378 - val_vae_kl_loss: 2.3842e-07`

Has anyone changed anything, or what are your experiences?

**Comment by Faur** [comment]
It looks like the posterior is collapsing

**Comment by AliBaheri** [comment]
Any update for this issue?

**Comment by Faur** [comment]
I ended up using this repo as a starting point instead
https://github.com/dariocazzani/World-Models-TensorFlow

**Comment by jneeven** [comment]
I initially had the same problem. I then looked at the actual code used by David Ha himself (https://github.com/hardmaru/WorldModelsExperiments) and discovered that they use slightly different loss functions.

This Keras version multiplies the vae_r_loss by 10 for some reason, and both loss functions compute the mean over all dimensions. By contrast, the actual World Models code sums the loss over all dimensions, and then averages per batch. Additionally, they seem to use a maximum function to make sure that the KL loss is never below a given value (KL_TOLERANCE * Z_DIM). In their code, KL_TOLERANCE is 0.5.

The Keras implementation of the World Models functions is as follows:
```
        def vae_r_loss(y_true, y_pred):
            loss = K.sum(K.square(y_true - y_pred), axis = [1, 2, 3])
            return K.mean(loss)

        def vae_kl_loss(y_true, y_pred):
            loss = - 0.5 * K.sum(1 + vae_z_log_var - K.square(vae_z_mean) - K.exp(vae_z_log_var), axis = 1)
            loss = K.maximum(loss, KL_TOLERANCE * Z_DIM)
            return K.mean(loss)
```

Replacing these functions in vae/arch.py resulted in pretty nice reconstructions for me.

**Comment by davidADSP** [comment]
Yes the reconstruction loss needs to be multiplied to ensure a balance between this and the KL divergence. I have updated the codebase to run with TF2.2 now and included this factor.
